ltdconveyor/s3upload.py

```diff
@@ -0,0 +1,395 @@
+        in the ``x-amz-meta-surrogate-key`` field. This key is used to purge
+        builds from the Fastly CDN when Editions change.
+        If `None` then no header will be set.
+    cache_control : `str`, optional
+        This sets the ``Cache-Control`` header on the uploaded
+        files. The ``Cache-Control`` header specifically dictates how content
+        is cached by the browser (if ``surrogate_control`` is also set).
+    surrogate_control : `str`, optional
+        This sets the ``x-amz-meta-surrogate-control`` header
+        on the uploaded files. The ``Surrogate-Control``
+        or ``x-amz-meta-surrogate-control`` header is used in priority by
+        Fastly to givern it's caching. This caching policy is *not* passed
+        to the browser.
+    acl : `str`, optional
+        The pre-canned AWS access control list to apply to this upload.
+        Defaults to ``'public-read'``, which allow files to be downloaded
+        over HTTP by the public. See
+        https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl
+        for an overview of S3's pre-canned ACL lists. Note that ACL settings
+        are not validated locally.
+    aws_access_key_id : `str`, optional
+        The access key for your AWS account. Also set
+        ``aws_secret_access_key``.
+    aws_secret_access_key : `str`, optional
+        The secret key for your AWS account.
+    aws_profile : `str`, optional
+        Name of AWS profile in :file:`~/.aws/credentials`. Use this instead
+        of ``aws_access_key_id`` and ``aws_secret_access_key`` for file-based
+        credentials.
+
+    Notes
+    -----
+    ``cache_control`` and  ``surrogate_control`` can be used together.
+    ``surrogate_control`` takes priority in setting Fastly's POP caching,
+    while ``cache_control`` then sets the browser's caching. For example:
+
+    - ``cache_control='no-cache'``
+    - ``surrogate_control='max-age=31536000'``
+
+    together will ensure that the browser always does an ETAG server query,
+    but that Fastly will cache the content for one year (or until purged).
+    This configuration is good for files that are frequently changed in place.
+
+    For immutable uploads simply using ``cache_control`` is more efficient
+    since it allows the browser to also locally cache content.
+
+    .. seelso:
+
+       - `Fastly: Cache control tutorial
+         <https://docs.fastly.com/guides/tutorials/cache-control-tutorial>`_.
+       - `Google: HTTP caching <http://ls.st/39v>`_.
+    """
+    log.debug('s3upload.upload({0}, {1}, {2})'.format(
+        bucket_name, path_prefix, source_dir))
+
+    session = boto3.session.Session(
+        profile_name=aws_profile,
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key)
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(bucket_name)
+
+    metadata = None
+    if surrogate_key is not None:
+        if metadata is None:
+            metadata = {}
```

Comment:
I've pushed checking whether metadata is empty to `upload_file` and `upload_object`.